jobs/batch_jobs/opportunity_jobs.py

```diff
@@ -0,0 +1,188 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jobs that are related to opportunity models.."""
+
+from __future__ import absolute_import
+from __future__ import annotations
+from __future__ import unicode_literals
+
+from core.domain import exp_fetchers
+from core.domain import opportunity_services
+from core.domain import story_fetchers
+from core.domain import topic_fetchers
+from core.platform import models
+from jobs import base_jobs
+from jobs.io import ndb_io
+from jobs.types import job_run_result
+
+import apache_beam as beam
+
+from typing import List, cast # isort:skip
+
+MYPY = False
+if MYPY:
+    from mypy_imports import datastore_services
+    from mypy_imports import exp_models
+
+
+(exp_models, opportunity_models, 
+topic_models, story_models,) = (
+    models.Registry.import_models([
+        models.NAMES.exploration, models.NAMES.opportunity,
+        models.NAMES.topic, models.NAMES.story]))
+
+
+class ExplorationOpportunitySummaryModelRegenerationJob(base_jobs.JobBase):
+    """Job that regenerates ExplorationOpportunitySummaryModel."""
+
+    @staticmethod
+    def _regenerate_opportunities_related_to_topic(topic, stories_dict, exploration_dict):
+        try:
+            story_ids = topic.get_canonical_story_ids()
+            stories = [stories_dict.get(id, None) for id in story_ids]
+            exp_ids = []
+            non_existing_story_ids = []
+
+            for index, story in enumerate(stories):
+                if story is None:
+                    non_existing_story_ids.append(story_ids[index])
+                else:
+                    exp_ids += story.story_contents.get_all_linked_exp_ids()
+            
+            exp_ids_to_exp = {
+                eid: exploration_dict[eid] for eid in exp_ids 
+                if exploration_dict.get(eid) is not None
+            }
+            non_existing_exp_ids = set(exp_ids) - set(exp_ids_to_exp.keys())
+
+            if len(non_existing_exp_ids) > 0 or len(non_existing_story_ids) > 0:
+                raise Exception(
```

Comment:
Should we create a specific exception here co that we catch only that exception.